File: Java/ds/SegmentTree.java
/*Implementation of Segment Tree for a problem where
	we have to find the sum of element from a given range(l,r) 
and also perform the update operation at particular index/*
*/

public class SegmentTree{
	
//	A function for returning the size of the segment Tree
	public static int sizeOfSegmentTree(int n) {
		 
		 int y = (int) (Math.ceil(Math.log(n) / Math.log(2))); 
		 int size = 2 * (int) Math.pow(2, y) - 1; 
	  
	  return size;
	}
	
	
//	A recursive function for creating a segment tree from a given array	
	public static void buildTree(int start,int end,int treeIndex,int arr[],int tree[]) {

		if(start==end) {
			tree[treeIndex]=arr[start];
			return;
		}
		
		int mid=(start+end)/2;
		buildTree(start, mid, 2*treeIndex, arr, tree);
		buildTree(mid+1, end, 2*treeIndex+1, arr, tree);
		
		tree[treeIndex]=tree[2*treeIndex]+tree[2*treeIndex+1];
	}
	
	
//	Function for updating a value at particular index	
	public static void update(int start,int end,int index,int value,int treeIndex,int arr[],int tree[]) {
		
		if(start==end) {
			arr[start]=value;
			tree[treeIndex]=value;
			return;
		}
		
		int mid=(start+end)/2;
		if(index>mid) {
//			right side tree
			update(mid+1, end, index, value, 2*treeIndex+1, arr, tree);
		}else {
//			left side tree
			update(start, mid, index, value, 2*treeIndex, arr, tree);
		}
		
		tree[treeIndex]=tree[2*treeIndex]+tree[2*treeIndex+1];
	}
	
//	Function which return the sum of elements from left index to right index	
	public static int query(int start,int end,int left,int right,int treeIndex,int tree[]) {
		
//		Completely outside
		if(start>right||end<left) {
			return 0;
		}
//		Completely inside
		if(start>=left&&end<=right) {
			return tree[treeIndex];
		}
//		Partial 
		int mid=(start+end)/2;
		
		int leftSum=query(start, mid, left, right, 2*treeIndex, tree);
		int rightSum=query(mid+1, end, left, right, 2*treeIndex+1, tree);
		
		return (leftSum+rightSum);
	}
	
	public static void main(String[] args) {
		
		int arr[]= {1,3,2,4,1,5};
		int n=arr.length;
		
		int size=sizeOfSegmentTree(n);
		
//		Segment Tree
		int tree[]=new int[size];
		buildTree(0, n-1, 1, arr, tree);
		
		System.out.println("Sum of element from 2 to 5 index : "+query(0, n-1, 2, 5, 1, tree));
		
//		Updating value at index 3 in given array to 6
		int index=3;
		int value=6;
		update(0, n-1, index, value, 1, arr, tree);
//		After updation new Array : {1,3,2,6,1,5}
		
		
		System.out.println("Sum of element from 2 to 5 index : "+query(0, n-1, 2, 5, 1, tree));
		
	}

}

File: Python/other/Segment_Tree.py
"""
Purpose: Let us consider the following problem to understand Segment Trees.
        We have an array arr[0 . . . n-1]. We should be able to 
        1 Find the min of elements from index l to r where 0 <= l <= r <= n-1
        2 Change value of a specified element of the array to a new value x. 
        We need to do arr[i] = x where 0 <= i <= n-1.
         
Article Link:- https://www.geeksforgeeks.org/segment-tree-set-1-sum-of-given-range/

Method: Segment Tree
Intution: Like Heap, the segment tree is also represented as an array. 
        The difference here is, it is not a complete binary tree. It is
        rather a full binary tree (every node has 0 or 2 children) and all
        levels are filled except possibly the last level. Unlike Heap,
        the last level may have gaps between nodes.

Time Complexity:  
        O(log(n)) -> for finding the min
        O(log(n)) -> for update query
Space Complexity: O(N)

Argument: Array containg the elements
Return  : for each query, return an integer denoting the desired ans

"""

from time import time
class Seg:
    def __init__(self,arr):
        self.size=1
        n = len(arr)

        # intilising the Segment Tree
        while(self.size<n):
            self.size*=2
        self.seg=[10e9]*(2*self.size)

        for i in range(n):
            self.seg[self.size-1+i]=arr[i]
        temp=self.size//2-1

        while(temp>=0):
            for i in range(temp,2*temp+1):
                self.seg[i]=min(self.seg[2*i+1],self.seg[2*i+2])
            temp=(temp+1)//2-1

    # For update query
    def update(self,x,y):
        pos=self.size-1+x
        self.seg[pos]=y
        temp=(pos-1)//2
        while(temp>=0):
            self.seg[temp]=min(self.seg[2*temp+1],self.seg[2*temp+2])
            temp=(temp-1)//2

    # for fing query        
    def find(self,x,lx,rx,l,r):
        if(rx<=l) or (lx>=r):
            return 10e9
        if(l<=lx and rx<=r):
            return self.seg[x]
        ans=0;mid=(lx+rx)//2
        ans+=min(self.find(2*x+1,lx,mid,l,r),self.find(2*x+2,mid,rx,l,r))
        return ans

# ------------------------DRIVER CODE ------------------------

if __name__ == "__main__":

    # Input the initial Position of the Knight
    n = int(input("Enter the length of array "))
    print("Enter the Array ")
    arr = [int(i) for i in input().split()]
    query = int(input("Enter the nuber of queries "))
    start = time()
    seg_tree = Seg(arr)
    
    # for each query will consist of 3 integers
    # <query type> <index> <index/val>
    for _ in range(query):
        query_type, x, y = map(int, input().split())
        
        # for update Query
        if query_type == 1:
            seg_tree.update(x,y)
            arr[x-1] = y
            print("Value at index ", x, " is updated to", y)
            print("Array: ", *arr)
        
        elif query_type == 2:
            print("Minimum from the given range is ", seg_tree.find(0,0,seg_tree.size,x,y))
        

    print("Time taken: ", time()-start)

"""
Sample Input / Output

Enter the length of array 6
Enter the Array
2 4 6 3 3 2
Enter the nuber of queries 4
2 3 6
Minimum from the given range is  2
1 6 5
Value at index  6  is updated to 5
Array:  2 4 6 3 3 5
2 3 6
Minimum from the given range is  3
2 5 5
Minimum from the given range is  3
Time taken:  65.34857654571533
"""
